### PR TITLE
Cache size added as last argument (conditional)

### PR DESCRIPTION
**Pull request by bratushka**

### CODE REVIEW

**Comment by toomuchdesign** [comment]
Hi @bratushka,
thanks for your PR. It seems cache control is a quite requested feature!

I checked your PR. It brings some interesting ideas, but there are a few implementation details I'd like to evaluate.

I think `cacheSize` param after `reselect`'s args makes the interface dirtier.

I also find the cache purge implementation a bit opinionated. What If I wanted to clear multiple cache items when I hit the cache size? Or remove them all as [Lodash does](https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L6461)? Or make one on my own?

I'd like to get to a simpler solution or (disclaimer: free ideas flowing around) something like an external cache management wrapper to plug to the selector creator on init phase.

**Idea 1:** add a new method to call during runtime to clear cache if length is larger then `X`.
```
reReselectInstance.clearCacheIfLengthExceeds(`cacheLength`);
```
**Idea 2:** expose a named configurable High Order Function (eg. `withCacheLimit`), to enhance the original re-reselect selector with caching management capabilities. The HOF should be able to use original re-reselect API's to clear cache, too. I'm not sure about the feasibility, but I like this approach.
```
import createCachedSelector, { withCacheLimit } from "re-reselect";

const createLimitedCachedSelector = withCacheLimit(createCachedSelector, cacheLength, cacheClearingFunction);

// Keep same API
createLimitedCachedSelector(...)(...);
```
I've probably written a lot of crap, I'll think about it. Any feedback, @bratushka ?
Cheers!

**Comment by bratushka** [comment]
Hello @toomuchdesign ,
Thanks for such a fast reaction, didn't expect it so early!

**1.** Talking about more dirty interface. The interface in my PR is made in a way that:

a) It 100% supports your current notation, so there's no need to fix anything in previously written code if you accept the PR. Reason: `cacheSize` is identified by its type and position, not just by position.

b) You need only 1 command to create the re-reselector. In my opinion 
```
const selector = createCachedSelector(argFunc, 10)(resultFunc);
```
 is nicer than
```
const selector = createCachedSelector(argFunc)(resultFunc);
selector.limitCacheTo(10);
```

If you disagree - we can still discuss on this.

**2.** Talking about removing multiple cache items or clear cache like in lodash.

When developing this piece of functionality I was inspired by the [Python's `functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache). 

Reason why I don't think it has sense to remove the cache completely or in chunks bigger than 1 item: **the more cache you have - the faster computation is.**

Reason why I still think sometimes you need to limit cache size: **memory**. One of projects I'm working at is an internal chat. Messages are encrypted. Each of them is about 1KB. I get them from redux store using a selector. Here's calculation of how much memory an unlimited reselector would require for a chat which goes from 0 to 200 messages: 200 (quantity of messages - so quantity of cached selectors) * (200/2) (medium quantity of messages in one selector) * 1KB (size of each message) = 20MB (just for 1 chat in application with multiple chats).

**3.** Talking about a higher order function: a factory of reselectors. Don't you think it's a bit too much abstraction? I wanted to keep it as simple as possible, 100% following what you've done.

Waiting to hear back from you,
Cheers!

**Comment by xsburg** [comment]
Hello,
Let me give you my two cents on what's been discussed so far.

@toomuchdesign, HOF what you've suggested in the second example is basically the decorator pattern. It means that you can only operate with the existing API of `createCachedSelector` function which is not enough to manage the cache behaviour in the way the HOF needs. Therefore it would mean introducing an option in  `createCachedSelector` to handle cache behaviour, something like the `memoize` argument in `reselect` library. At the moment, I can't see use cases to support this increase in complexity and would rather go for a simpler version.

Having said that, I don't think we should add any arguments in the original function that basically copies the `reselect` API now, which is really neat. Instead, an additional argument can be added just like the  `selectorCreator` parameter.

Here's what I have in mind:
```
createCachedSelector(<...reselect arguments>)(keyFn, { selectorCreator, cacheSize }|selectorCreator|undefined);
```

We preserve backward compatibility by still accepting `selectorCreator` as the second parameter, but from now on it's an options object, which adds verbosity to the magic numbers of the cache size.

What do you make of this?

@bratushka, I've been trying to understand the maths in your example but I still can't see why 200 messages of 1KB in size transform into 20MB of data. Given that you have one cached selector instance per chat it means that you only double the data (which is state data + cached data) so the memory consumed is N rather N^2. What am I missing?

And regarding the implementation, I'd rather put the extra logic into `if (cacheSize) { ... }` guard, so that we don't update the `cacheOrdering` array unless the flag is on.

**Comment by toomuchdesign** [comment]
Hi @xsburg,
Your cents are very welcome!

Green light on introducing a **configuration object**. I had in mind the same idea for a major release. For now we can accept both a `selectorCreator` and the config object and restrict the API to the config object in the next major release. 👍 

I still have a few doubts about the cache clearing logic. In my experience `reselect` is not only important for evaluation speed, but mostly for **stabilizing state derivation**.

```
const state = [1, 2, 3, 4, 5];

// Without reselect
const firstThreeEntries = getFirstThreeEntries(state); // [1, 2, 3]
```
Without `reselect` or any sort of memoization,`getFirstThreeEntries` returns a new identical array on each call, forcing React to re-evaluate all those UI components basing updating strategy on on props identity equality (aka shallow equality).

For example, I wrote this library for a project with a lot of animated data/charts visualization: each re-rendering made the charts animate again from the beginning, so having a stable state derivation became vital.

That said, **removing the oldest cache item** is a good default, but I suspect It might not fit a lot of use cases. How do you feel about this?

How would it be it we added an **optional `cacheClearingFunction` config option** accepting a function with defined parameters to let users write its own logic?
```
{
  cacheClearingFunction: (cache, cacheOrdering) => {
    // Mutate `cache` and  `cacheOrdering`?
  }
}
```

Thanks you all for your contribution!

**Comment by bratushka** [comment]
Hello @toomuchdesign @xsburg ,
Thanks for your activity, I appreciate that very much.
Why don't we make a decision on how the API should look like, so everybody will be happy? Internal realization is not that important (considering it works as expected). I think we all agree that the thing we should concentrate on is how handy it will be to use the library.

So my proposal in this PR is the following:
```
const selector = createSelector(...inputFuncs, outputFunc)(cacheKeyFunc)  // Without cacheSize
const selector = createSelector(...inputFuncs, outputFunc, 10)(cacheKeyFunc)  // cacheSize limited to 10 results
```
This is all you need. No wrappers, no configs, 100% backward compatibility.

What are your proposals? I'd be happy to refactor the PR if you have some nice and handy idea.

Cheers guys!

P.S.: explanation of calculation of 20MB cache size.
**1)** We started with 1 message, then one by one we increase the quantity of messages to 200. On every new message we reselect all messages (imagine you select them as a list of messages, not a list of IDs for whatever reason). This means you have 200 lists in cache. 200 cached results.
**2)** Medium quantity of messages in each list is `(1+2+3+...+199+200)/200` which is `100.5` (I've rounded it to 100 for nice calculations)
**3)** Each message is ~ 1KB

**Summary)** 
```
quantity_of_results * medium_length_of_result * size_of_message =
200 * 100 * 1KB =
20MB
```

**Comment by alexanderjarvis** [comment]
Just catching up on this thread as I may be interested a bounded cache.

Two points to consider:

1. An LRU cache would probably be more appropriate than just chucking the last item.
1. Being able to inject a `cache` for re-reselect where the object conforms to a protocol would be the most flexible solution. It only needs a handful of functions and there can be a default implementation provided by re-reselect.

**Comment by bratushka** [comment]
@alexanderjarvis now LRU rule is applied (reused result goes to the top of the cache). If you want to have a possibility to inject cache - please create a separate issue on it, I'll do that separately. I still believe that every ticket/branch should do one thing at a time :)

File: index.js [hunk 1]
```diff
@@ -20,6 +20,9 @@ export default function createCachedSelector(...funcs) {
             delete cache[earliest];
           }
         }
+
+        // Move the latest cache key to the top.
+        cacheOrdering = [...cacheOrdering.filter(key => key !== cacheKey), cacheKey];
```
Comment (alexanderjarvis) [inline index.js]: put this in an `else` directly above so that you're not doing this when you just used `push`.

**Comment by alexanderjarvis** [comment]
@bratushka looks good! just a simple comment about an unnecessary operation.

**Comment by bratushka** [comment]
@alexanderjarvis nice spot, fixed

**Comment by toomuchdesign** [comment]
Hi @bratushka, hi @alexanderjarvis,
thank you for your efforts! 👏 

As @alexanderjarvis ’s first comment outlined, cache clearing subject is more subtle than It appears at first sight. We are just 4 people and I can see 3 different ideas about how cache clearing should work! So let’s take our time to make the right decisions.

As the author of this library, I want to try to include this feature but also:

1. Guarantee a stable and **meaningful API** the the users
2. Find a flexible and dry way to **customise the cache clearing algorithm** (not picking up a random one)
3. Guarantee the same current **performances** to the users who don’t need to limit the cache
4. Pay attention to the package’s **final weight**

### 1.
As for the API, I’m with @xsburg proposal of adding an optional config object:

```
createCachedSelector(<...reselect arguments>)(keyFn, { selectorCreator, cacheSize }|selectorCreator|undefined);
```

### 2.
As for the custom cache clearing, things get tougher. The first solution that comes to my mind is to inject the caching logic by providing **2 different function as options** (with defaults). But It probably still wouldn’t cover all the cache use cases. **Any other idea?** 🤔 
```
{
   cacheLength,
   manageCacheOnNewEntry: (cacheKey, cacheKeys, cacheSize) => { … },
   manageCacheOnExistingEntry: (cacheKey, cacheKeys, cacheSize) => { … },
}
// PS. Naming still to be reviewed
```
Please consider that any future change on how cache clearing works, will probably result into a breaking change.

### 3.
Depending on what we do with point 2, we’ll decide whether including the feature into current re-reselect selector creator or expose it as a separate named export. For example, I’d like to not overload current users' code with cacheOrdering updates when they don’t need it.

If we incapsulate cache clearing logic into a couple of `if` conditions, instead, I suppose we would be able to avoid a separate export. (How do you see it @xsburg?)

Thank you all again!

**Comment by xsburg** [comment]
Thanks everybody, we are really getting somewhere with that.

Having resolved that we should extract the cache logic somehow, I think the best way to do that would be not a set of functions, but a strategy object. Consider that we introduce `cacheObject` option with the following interface:

```typescript
interface ICacheObject {
    add (key: string|number, selectorFn: Function): void;
    get (key: string|number): Function;
}
```

By this, it's now possible to implement both the trivial cache and the LRU-based one, here is a draft of it:

```javascript
class TrivialCacheObject {
    add (key, selectorFn) {
        this._cache[key] = selectorFn;
    }
    get (key) {
        return this._cache[key];
    }
}

class LruCacheObject {
    constructor ({ cacheSize }) {
        //...
    }
    add (key, selectorFn) {
        this._cache[key] = selectorFn;
        const index = this._cacheOrdering.indexOf(key);
        if (index) {
            this._cacheOrdering.splice(index, 1);
        }
        this._cacheOrdering.push(key);
        if (this._cacheOrdering.length > this._cacheSize) {
            this._cacheOrdering.shift();
        }
    }
    get (key) {
        return this._cache[key];
    }
}
```

By default, we just set the trivial cache object and the implementation is only one function call slower. In contrast, we have the flexibility we need to implement any type of caching algorithm.

@toomuchdesign, what do you think?

**Comment by toomuchdesign** [comment]
I read a few time the proposal and I think it is a very powerful approach. 😍 

I personally don't like classes in JS, but since they have been adopted as a standard by React itself, we are authorized to try this way.

I also would have preferred to solve this in a pure functional way, but I don't see how to get to the same result in terms of flexibility and customization.

### How does it work?
If I get it right, the user (optionally) supplies a `CustomCache` class as an option, and `re-reselect` create/uses an instance of that class like:

```
const cache = new options.CustomCache(options.cacheLength);
```
Then it's just a matter of calling cache objects methods.

#### Point 1:
The object oriented god, might claim to **extend an abstract `CustomCache`** class with basic methods/implementations. I personally find it excessive and avoid it since it would transpile into quite crappy code. Agree/disagree?

#### Point 2:
`.removeMatchingSelector` still remains uncovered by the proposed cache interface. In order to keep it, `clear` method should accept an optional `key`argument, or the cache object should declare a further `clearKey` method. Not nice.

While `.getMatchingSelector()` is necessary to test/debug original `reselect` selectors by accessing their methods, `.removeMatchingSelector()` was added to free the cache by manually removing some items.

**Should we remove `.removeMatchingSelector()` method** and go for a breaking change or try to preserve the full existent API? (Hi, @rufman !)

#### Point 3:
Just to not reinvent the wheel, we might consider to provide a few custom ready-to-use cache objects, or wrap some existing libraries. _The nice part is that, by extracting caching logic, we can test it in isolation.

#### Point 4:
(Just a detail): `LruCacheObject.get()` should also update `_cacheOrdering`, right?

### A few references (API naming and Point 3)
- https://www.npmjs.com/package/lru-cache
- https://www.npmjs.com/package/flat-cache
- https://www.npmjs.com/package/simple-cacher
- http://world.episerver.com/documentation/Items/Tech-Notes/EPiServer-CMS-5/EPiServer-CMS-5-SP1/Pluggable-Runtime-Cache/

Thank you @xsburg for giving all these ideas a quite solid shape!

**Comment by xsburg** [comment]
Thanks, @toomuchdesign!

I think I need to clarify how I see the API in my proposal just a little.
First, regarding classes, my idea is that we pass an instance of a cache object, not the class:

```javascript
// If we want to use one of the predefined algorithms
import createCachedSelector, { LruCacheObject } from 're-reselect';
createCachedSelector(...reselectArgs)(keySelector, {
    cacheObject: new LruCacheObject({ cacheSize: 1024 })
});
```

I find this a much better approach rather than passing a class for two reasons:

1. We don't have to pass any extra options into `createCachedSelector` (like `options.cacheLength` in your example).
2. We don't oblige the user to create a class, only to provide an object of the declared shape.

So somebody could meet the interface by providing the cache object like this:

```javascript
lru = LRU({ max: 1024 });
createCachedSelector(...reselectArgs)(keySelector, {
    // We can provide an object that we create on the fly, even though it might be a bit messy in this case.
    cacheObject: {
        add (key, value) {
            lru.set(key, value);
        },
        remove(key) {
            lru.del(key);
        }
        // ...
    }
});
```

Regarding the points that's been mentioned:

1. Abstract class for `cacheObject` - agreed, I don't actually see the point of creating a base class in this case, primarily because the implementation is so trivial.
2. I think it's fine to extend the interface by adding the `remove` or `del` method (like in the `lru-cache` package). After all, other cache implementations provide it.
3. Makes perfect sense, so far I can see that we provide the trivial and LRU-based versions of the cache. I've come up with the following options:
    - we write our own implementation of the LRU-cache
    - we add dependency to `lru-cache` and base our implementation on it
    - we adopt our naming to conform to `lru-cache` and refer to it in the docs
    
    I'm inclining toward writing our own implementation, so that we avoid excessive complexity and keep the performance high (by no means we need disposal callbacks, for example). Thoughts?
4. Of course, just a typo.

**Comment by toomuchdesign** [comment]
From my side we are ready to define the last details and start!

As for the cache API's, I like `lru-cache` naming, but I'd probably just borrow its`set` method. So I'd replace `add` with `set` and add a `remove` method.

```
set(key, value)    // we "set" an item since the cache would be able to replace an existing one
get(key)
remove(key)        // we don't not delete the entry but just remove it from the cache
clear()
```
And - of course - let's try to provide a **lightweight implementation** of a few **caching strategies**:
- `lru-cache` Last Recently Used
- `fifo-cache` First In First Out
- `flat-cache` or just `cache` as default cache

 _(Yes, I googled the names)_ 😎 

In the upcoming week I will have a few time to work on it, so let's coordinate us.

In order to not overdo, I propose to **break the work into 2 releases**:

1. introduction of cache object with default `flat-cache` (+ docs, + TS typings, +  tests)
2. addition of the optional caching strategies (+ tests)

As always I appreciate any comment/suggestion.
Good night and thank you again!

**Comment by toomuchdesign** [comment]
Discussion moved to #24.

**Pull request closed without merging**